### PR TITLE
Add jax compatible api

### PR DESCRIPTION
**Pull request by sky-2002**
This PR adds a JAX compatible API, refer issue #1027 

Looking forward for review and open to feedback(especially in writing tests).

### CODE REVIEW

File: outlines/processors/base_logits_processor.py [hunk 1]
```diff
@@ -20,6 +22,12 @@ def is_mlx_array_type(array_type):
     return issubclass(array_type, mx.array)
 
 
+def is_jax_array_type(array_type):
+    return array_type == jaxlib.xla_extension.ArrayImpl or isinstance(
+        array_type, jaxlib.xla_extension.ArrayImpl
+    )
```
Comment (lapp0) [inline outlines/processors/base_logits_processor.py:28]: Please follow the pattern of `is_mlx_array_type`, where we first check if it's importable, since it's possible jax isn't installed.

File: tests/processors/test_base_processor.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+
+# ---- Test _to_torch() ---- #
+def test_to_torch_with_numpy(processor):
+    np_array = np.array([[1, 2], [3, 4]], dtype=np.float32)
+    torch_tensor = processor._to_torch(np_array)
+    assert isinstance(torch_tensor, torch.Tensor)
+    assert torch.allclose(
+        torch_tensor, torch.tensor([[1, 2], [3, 4]], dtype=torch.float32)
+    )
```
Comment (lapp0) [inline tests/processors/test_base_processor.py:46]: Glad to see you have good test coverage.

This could be made much simpler with 

```
arrays = [
    get_list_array(),
    get_np_array(),
    ...
]

@pytest.mark.parameterize("array", arrays)
def test_to_torch(array, processor):
    ...

**Review by lapp0 — CHANGES_REQUESTED** [review]
Great work so far!

Could you please add jax to `benchmarks/bench_processors.py` as well?

Also could you fill me in on how you intend to use a jax based model, just so I have context on how Outlines is being used here?

Thanks!

File: tests/processors/test_base_processor.py (hunk 2 above)
Comment (sky-2002) [reply tests/processors/test_base_processor.py]: Makes sense, done

**Comment by sky-2002** [comment]
Hey @lapp0 , thanks for the feedback.
Have added to benchmarks(though I haven't run those yet, will possibly do)

Regarding intent to use with JAX, @borisdayma(author of issue mentioned above) can help us understand jax based model usage with `outlines`.

PS: I currently have no use case with JAX, but I was trying out `outlines` and found it very interesting, so wanted to contribute.

File: tests/processors/test_base_processor.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+from outlines.processors.base_logits_processor import OutlinesLogitsProcessor
+
+arrays = {
+    "list": [[1.0, 2.0], [3.0, 4.0]],
+    "np": np.array([[1, 2], [3, 4]], dtype=np.float32),
+    "jax": jnp.array([[1, 2], [3, 4]], dtype=jnp.float32),
+    "torch": torch.tensor([[1, 2], [3, 4]], dtype=torch.float32),
+    "mlx": mx.array([[1, 2], [3, 4]], dtype=mx.float32),
+}
```
Comment (lapp0) [inline tests/processors/test_base_processor.py:24]:
```
try:
    import mlx.core as mx
    arrays["mlx"] = mx.array([[1, 2], [3, 4]], dtype=mx.float32)
except ImportError:
    pass
```

same for jax

File: tests/processors/test_base_processor.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import torch
+
+# Import mlx and the processor class
+try:
+    import mlx.core as mx
+
+    MLX_AVAILABLE = True
+except ImportError:
+    MLX_AVAILABLE = False
+
```
Comment (lapp0) [inline tests/processors/test_base_processor.py:15]: no need for this if you use the below comment

File: outlines/processors/base_logits_processor.py [hunk 5]
```diff
@@ -129,6 +145,11 @@ def _from_torch(tensor: torch.Tensor, target_type: Type) -> Array:
             # numpy doesn't support bfloat16, mlx doesn't support direct conversion from torch
             return mx.array(tensor.float().numpy())
 
+        elif is_jax_array_type(target_type):
+            import jax
+
+            return jax.numpy.from_dlpack(tensor)
```
Comment (lapp0) [inline outlines/processors/base_logits_processor.py:151]: I'm wondering if we need to use `jax.numpy`? Is there a reason to not just use `jax.dlpack.from_dlpack`?

File: outlines/processors/base_logits_processor.py [hunk 6]
```diff
@@ -101,6 +111,12 @@ def _to_torch(tensor_like: Array) -> torch.Tensor:
             # https://ml-explore.github.io/mlx/build/html/usage/numpy.html
             return torch.from_dlpack(tensor_like)
 
+        elif is_jax_array_type(type(tensor_like)):
+            import jax
+
+            torch_tensor = torch.from_dlpack(jax.dlpack.to_dlpack(tensor_like))
+            return torch_tensor.cuda()
```
Comment (lapp0) [inline outlines/processors/base_logits_processor.py:118]: We shouldn't send to cuda.

**Review by lapp0 — CHANGES_REQUESTED** [review]
>PS: I currently have no use case with JAX, but I was trying out outlines and found it very interesting, so wanted to contribute.

Thanks for clarifying and your interest in the project.

>Have added to benchmarks(though I haven't run those yet, will possibly do)

I'll add the `run-benchmarks` label, which will trigger their run once the workflow is approved.

>Regarding intent to use with JAX, @borisdayma(author of issue mentioned above) can help us understand jax based model usage with outlines.

@borisdayma could you review and smoke test to ensure this fits your desired use case?

File: outlines/processors/base_logits_processor.py (hunk 5 above)
Comment (sky-2002) [reply outlines/processors/base_logits_processor.py:151]: you are right, we can just use `jax.dlpack.from_dlpack` , idk why autocomplete suggestions didn't show it, and I came across links that used `jax.numpy.dlpack`, though I will check whats difference it creates really

File: outlines/processors/base_logits_processor.py (hunk 6 above)
Comment (sky-2002) [reply outlines/processors/base_logits_processor.py]: Fixed

File: outlines/processors/base_logits_processor.py (hunk 1 above)
Comment (sky-2002) [reply outlines/processors/base_logits_processor.py:30]: Done

File: tests/processors/test_base_processor.py (hunk 3 above)
Comment (sky-2002) [reply tests/processors/test_base_processor.py:15]: Done
